File: java-to-js/src/checked-exceptions.ts
class IOException extends Error {
  constructor(message: string) {
    super(message);
    this.name = "IOException";
  }
}

function main(): void {
  try {
    someIoOperation();
    parentIoOperation();
  } catch (e) {
    if (e instanceof IOException) {
      console.log(`Caught IOException: ${e.message}`);
    } else {
      throw e;
    }
  }

  someRuntimeErrorOperation();
}

main();

function parentIoOperation(): void {
  someIoOperation();
}

function someIoOperation(): void {
  console.log("Performing IO operation");
  throw new IOException("IO error occurred");
}

function someRuntimeErrorOperation(): void {
  console.log("Performing operation");
  throw new Error("Runtime error occurred");
}

File: java-to-js/src/CheckedExceptions.java
import java.io.IOException;

public void main() {
  try {
    someIoOperation();
    parentIoOperation();
  } catch (IOException e) {
    System.out.println(STR."Caught IOException: \{e.getMessage()}");
  }
 
  someRuntimeErrorOperation();
}

void parentIoOperation() throws IOException {
  someIoOperation();
}

void someIoOperation() throws IOException {
  System.out.println("Performing IO operation");
  throw new IOException("IO error occurred");
}

void someRuntimeErrorOperation() {
  System.out.println("Performing operation");
  throw new RuntimeException("Runtime error occurred");
}